File: tests/human/x/clj/group_by_conditional_sum.clj
(ns group-by-conditional-sum)

(def items
  [{:cat "a" :val 10 :flag true}
   {:cat "a" :val 5 :flag false}
   {:cat "b" :val 20 :flag true}])

(def result
  (->> items
       (group-by :cat)
       (map (fn [[cat xs]]
              (let [num (reduce + (map (fn [x] (if (:flag x) (:val x) 0)) xs))
                    den (reduce + (map :val xs))]
                {:cat cat
                 :share (/ num den)})))
       (sort-by :cat)))

(println result)

File: tests/human/x/ex/group_by_conditional_sum.ex
items = [
  %{cat: "a", val: 10, flag: true},
  %{cat: "a", val: 5, flag: false},
  %{cat: "b", val: 20, flag: true}
]

result =
  items
  |> Enum.group_by(& &1.cat)
  |> Enum.map(fn {cat, group} ->
    total = Enum.sum(Enum.map(group, & &1.val))
    flagged =
      group
      |> Enum.filter(& &1.flag)
      |> Enum.sum(Enum.map(& &1.val))
    %{cat: cat, share: flagged / total}
  end)
  |> Enum.sort_by(& &1.cat)

IO.inspect(result)
